Reduce time it takes to import SGLang

raayandhar opened the pull request:
<!-- Thank you for your contribution! Please follow these guidelines to enhance your pull request. If anything is unclear, submit your PR and reach out to maintainers for assistance. Join our Slack community at https://slack.sglang.ai to discuss further. -->

## Motivation

<!-- Describe the purpose and goals of this pull request. -->

We notice that the time to import things in SGLang takes a lot of time (https://github.com/sgl-project/sglang/issues/10492). I have been looking into what is taking up a lot of time and if there are simple ways to help reduce this import time. From the original issue, we want to reduce:
```
time python -c "from sglang.srt.managers.scheduler import Scheduler"
```
which is what I have been focusing my efforts on. However, I think there are things we can do to reduce time for other imports. This is more of a V1 to get community feedback from experts.

## Modifications

<!-- Detail the changes made in this pull request. -->

There are some heavy imports.  For example, the quantization methods import at the module level is heavy. Moving some imports to the function level (only time it is used), we can reduce module import time. However, I can see how this can easily be an antipattern. In fact, it can hurt performance if we have a function that is used a lot that we have an import in. I tried to only do this in functions that we only expect to run once or a small number of times. However, I can understand the argument against this kind of code. I also don't think all the changes to `hf_transformer_utils.py` help so I will be taken a deeper look, since the changes are a bit invasive.

## Accuracy Tests

These changes should not affect model outputs.

<!-- If this pull request affects model outputs (e.g., changes to the kernel or model forward code), provide accuracy test results. -->

## Benchmarking and Profiling

Running `for i in {1..100}; do (time python -B -c "from sglang.srt.managers.scheduler import Scheduler") 2>&1 | grep "^real"; done | python calc_avg.py` ([calc_avg.py](https://gist.github.com/raayandhar/ed637053e7c311630ed8e8cd989ce947))

With these changes:
```
===========Timing Statistics============                                                                            
Number of runs: 100                                                                                                 
Mean:     8.308s                                                                                                    
Median:   8.236s                                                                                                    
Std Dev:  0.297s                                                                                                    
Min:      7.999s                                                                                                    
Max:      9.329s                                                                                                    
========================================
```
Compared to top-of-main:
```
===========Timing Statistics============                                                                            
Number of runs: 100                                                                                                 
Mean:     9.836s                                                                                                    
Median:   9.790s                                                                                                    
Std Dev:  0.332s                                                                                                    
Min:      8.655s                                                                                                    
Max:      11.801s                                                                                                   
======================================== 
```
so we have ~1.5 second improvement. Not the best, so I am going to keep working on it. I mostly targeted improving the timing in the creation of `ModelConfig` object. The difference so far is largely from removing the quantization import:

Without these changes [import_sglang_tom.log](https://gist.github.com/raayandhar/d25a90632e3502f9f8791db333ad0b02)
<img width="271" height="166" alt="Screenshot 2025-11-01 at 9 03 47 PM" src="https://github.com/user-attachments/assets/aa476a4c-f9a6-4176-9b54-e9ebf2ea8b7b" />

With these changes [import_sglang_new.log](https://gist.github.com/raayandhar/0d851737cf7eae3218fca29c8607384a)
<img width="260" height="145" alt="Screenshot 2025-11-01 at 9 04 04 PM" src="https://github.com/user-attachments/assets/941b6174-e548-4e36-ae64-4ed92a30ade5" />

Machine:
- AMD EPYC 7343 16-Core Processor
- L40S GPU

## Checklist

- [x] Format your code according to the [Format code with pre-commit](https://docs.sglang.ai/developer_guide/contribution_guide.html#format-code-with-pre-commit).
- [x] Add unit tests according to the [Run and add unit tests](https://docs.sglang.ai/developer_guide/contribution_guide.html#run-and-add-unit-tests). (N/A?)
- [x] Update documentation according to [Write documentations](https://docs.sglang.ai/developer_guide/contribution_guide.html#write-documentations). (N/A?)
- [x] Provide accuracy and speed benchmark results according to [Test the accuracy](https://docs.sglang.ai/developer_guide/contribution_guide.html#test-the-accuracy) and [Benchmark the speed](https://docs.sglang.ai/developer_guide/contribution_guide.html#benchmark-the-speed).

raayandhar committed:
efforts done towards reducing import time

Signed-off-by: Raayan Dhar raayan.dhar@gmail.com <raayan.dhar@gmail.com>

raayandhar committed:
remove comment

Signed-off-by: Raayan Dhar raayan.dhar@gmail.com <raayan.dhar@gmail.com>

raayandhar commented:
I will continue working on this, there is more improvements to be made.

raayandhar commented on python/sglang/srt/utils/hf_transformers_utils.py:
```diff
@@ -13,78 +13,83 @@
-for name, cls in _CONFIG_REGISTRY.items():
-    with contextlib.suppress(ValueError):
-        AutoConfig.register(name, cls)
+def _register_custom_configs():
```
I think that a lot of these changes in this file do not really improve the times, and they are overly invasive. However, the transformers related code is a huge time increase. I think they may be impossible to remove since we init many transformer related objects in the scheduler.

raayandhar committed:
moving configs into _register() improves time

Signed-off-by: Raayan Dhar raayan.dhar@gmail.com <raayan.dhar@gmail.com>

raayandhar committed:
bigger improvements, reaching 7.5s-8s

Signed-off-by: Raayan Dhar raayan.dhar@gmail.com <raayan.dhar@gmail.com>

raayandhar commented on python/sglang/srt/disaggregation/decode.py:
```diff
@@ -296,6 +296,8 @@ def _init_kv_manager(self) -> BaseKVManager:
 
     def add(self, req: Req, is_retracted: bool = False) -> None:
         """Add a request to the pending queue."""
+        from sglang.srt.managers.schedule_batch import RequestStage
```
This is a very repetitive pattern, but the `from sglang.srt.managers.schedule_batch` import is somewhat expensive. The `RequestStage` object is just an enum, so we don't want to be doing the massive import for just an enum in my opinion. Maybe we can put this enum in a different file / some other solution that effectively does the same thing, so we can just import at the top level.

raayandhar replied:
I have refactored this into a separate file for just the request types so as to avoid some cognitive load. I can revert the changes if needed.

raayandhar committed:
more improvements, down to 3900 ms

Signed-off-by: Raayan Dhar raayan.dhar@gmail.com <raayan.dhar@gmail.com>

raayandhar commented on python/sglang/srt/layers/moe/__init__.py:
```diff
@@ -12,6 +11,10 @@
 )
+from sglang.utils import LazyImport
+
+MoeRunner = LazyImport("sglang.srt.layers.moe.moe_runner.runner", "MoeRunner")
```
I think this (`LazyImport(...)`)is something we can apply to many other files as well, I'm not sure if there's a downside?

zhyncs commented:
The drawback is that it introduces some extra cognitive load. May you fix the conflicts first? Thanks!

raayandhar committed:
forgot to run precommit

Signed-off-by: Raayan Dhar raayan.dhar@gmail.com <raayan.dhar@gmail.com>

raayandhar committed:
fix decode

Signed-off-by: Raayan Dhar raayan.dhar@gmail.com <raayan.dhar@gmail.com>

raayandhar committed:
move to lazy import

Signed-off-by: Raayan Dhar raayan.dhar@gmail.com <raayan.dhar@gmail.com>

raayandhar committed:
refactor: move manager request types to separate file to avoid antipattern

Signed-off-by: Raayan Dhar raayan.dhar@gmail.com <raayan.dhar@gmail.com>

raayandhar committed:
small tiny changes

Signed-off-by: Raayan Dhar raayan.dhar@gmail.com <raayan.dhar@gmail.com>

raayandhar committed:
small cleanup

Signed-off-by: Raayan Dhar raayan.dhar@gmail.com <raayan.dhar@gmail.com>

raayandhar commented on python/sglang/srt/disaggregation/decode.py:
```diff
@@ -884,6 +885,8 @@ def get_next_disagg_decode_batch_to_run(
 
     def get_new_prebuilt_batch(self: Scheduler) -> Optional[ScheduleBatch]:
         """Create a schedulebatch for fake completed prefill"""
+        from sglang.srt.managers.schedule_batch import ScheduleBatch
```
iiuc, this is because we import it at the top for the type checking, and then here we use it (`ScheduleBatch.init_new`, only function in this file that doesn't use it as a type) so we need to import it for use in this function. But I can leave it at top-level import if you think that's better and remove from type checking.

raayandhar commented:
> Some principles:
> 
> 1. Do not import inside a Python code block (function, class definition, if block...); import at the top of a file in most cases.
> 2. If an imported item works as a util function, you can import it when using it (inside a code block)
> 3. Place all type-hint-only class types under `if TYPE_CHECKING:`
> 
> @raayandhar

The worry I have is that some imports are only used in a single function across the entire file, and this function may also only be called once / very few times. For example, the `def _verify_quantization(self)` function is the only one that needs `QUANTIZATION_METHODS`, and importing this at the top level is very expensive. Maybe we can also consider these to be exception? Otherwise, we won't be reducing the time very much. Appreciate discussion on this matter. @hnyls2002

raayandhar committed:
get the correct imports

Signed-off-by: Raayan Dhar raayan.dhar@gmail.com <raayan.dhar@gmail.com>

raayandhar committed:
fix server_args

Signed-off-by: Raayan Dhar raayan.dhar@gmail.com <raayan.dhar@gmail.com>

raayandhar committed:
small fix to hf_transformer_utils

Signed-off-by: Raayan Dhar raayan.dhar@gmail.com <raayan.dhar@gmail.com>

raayandhar committed:
use lazyimport

Signed-off-by: Raayan Dhar raayan.dhar@gmail.com <raayan.dhar@gmail.com>

raayandhar committed:
handle OR and other checks

Signed-off-by: Raayan Dhar raayan.dhar@gmail.com <raayan.dhar@gmail.com>

raayandhar committed:
simplify

Signed-off-by: Raayan Dhar raayan.dhar@gmail.com <raayan.dhar@gmail.com>

raayandhar committed:
fix regression after rebase

Signed-off-by: Raayan Dhar raayan.dhar@gmail.com <raayan.dhar@gmail.com>

hnyls2002 commented:
Please fix the conflicts. @raayandhar 

Also I think this PR should be review by @merrymercy

raayandhar commented:
I investigated the previous CI errors and I think they be unrelated, to my best understanding, but I will monitor carefully again.
The changes are a bit extensive across a lot of different files. I think one strategy I think of is for when importing heavy classes, using `LazyImport` and keep it at top-level, to avoid cognitive load (and perhaps note that the import is expensive, so that is why we do it). When importing a util function with a heavy path, if it is just used once in the whole file, keep it at the function level. And then of course when only used for type checking, leave it under a`TYPE_CHECKING` block. I already do the latter two, but can update my PR to do the former. I think this is close in spirit to the earlier review. I think that might be the cleanest way to handle this. Happy to discuss further.

raayandhar committed:
refactor work

Signed-off-by: Raayan Dhar raayan.dhar@gmail.com <raayan.dhar@gmail.com>

raayandhar committed:
fix issue about import path

Signed-off-by: Raayan Dhar raayan.dhar@gmail.com <raayan.dhar@gmail.com>

raayandhar commented:
I've refactored the PR in the following way, based on earlier discussion. I liked the idea of having simple rules like @hnyls2002 mentioned:

> Some principles:
> 
> 1. Do not import inside a Python code block (function, class definition, if block...); import at the top of a file in most cases.
> 2. If an imported item works as a util function, you can import it when using it (inside a code block)
> 3. Place all type-hint-only class types under `if TYPE_CHECKING:`

I've refactored the PR so we do this. Now, in order to actually make the import times decrease, we need to avoid doing very expensive imports at import time. So I propose a fourth principle:

4. If a class import (i.e. non-util function) is very expensive, use `LazyImport` at the top/module level, use sparingly.

This helps us avoid the large import times. The only downside I see is that there is some cognitive load as mentioned earlier, but I think this is a better solution than importing at the function level. This also helps avoid some circular import issues, which could potentially be caused from the way I was doing it earlier.

Here are the most recent timings. On this branch (range 7-8.5 seconds):
```
(sglang) raayan@uril-1:~/projects/sglang$ time python -c "from sglang.srt.managers.scheduler import Scheduler"

real	0m7.469s
user	0m15.162s
sys	0m3.141s
```
On main (range anywhere from 10-15 seconds, usually 10.5-12 seconds):
```
(sglang) raayan@uril-1:~/projects/sglang$ time python -c "from sglang.srt.managers.scheduler import Scheduler"

real	0m10.981s
user	0m21.909s
sys	0m4.143s
```
At this point, we can't do much better. Looking at the import time heatmap, it is nearly all torch now, at 4.5 ms:
<details>
  <summary>this branch branch import time graph (click to expand)</summary>

<img width="3000" height="2000" alt="importtime_feat" src="https://github.com/user-attachments/assets/99fe304d-b9f4-4cd0-b790-c2a1e24d47ae" />

I recommend clicking on the image and expanding.
</details>

whereas main is at nearly 7 ms:
<details>
  <summary>main branch import time graph (click to expand)</summary>

<img width="3000" height="2000" alt="importtime_main" src="https://github.com/user-attachments/assets/5342e29a-8eb6-45a2-a3df-48da91a70922" />

I recommend clicking on the image and expanding.
</details>

which gives us a ~33% decrease.

raayandhar committed:
small fix

Signed-off-by: Raayan Dhar raayan.dhar@gmail.com <raayan.dhar@gmail.com>

raayandhar committed:
another small fix

Signed-off-by: Raayan Dhar raayan.dhar@gmail.com <raayan.dhar@gmail.com>

raayandhar committed:
small error with config

Signed-off-by: Raayan Dhar raayan.dhar@gmail.com <raayan.dhar@gmail.com>

raayandhar committed:
fix precommit

Signed-off-by: Raayan Dhar raayan.dhar@gmail.com <raayan.dhar@gmail.com>
